utilsForFrontend.mjs

```diff
@@ -459,30 +459,29 @@ export function sigmoid(x) {
 
 /**
  * Return whether an element is practically visible, considing things like 0
- * size or opacity, ``display: none``, and ``visibility: hidden``.
+ * size or opacity, and clickability.
  */
 export function isVisible(fnodeOrElement) {
     const element = toDomElement(fnodeOrElement);
-    for (const ancestor of ancestors(element)) {
-        const style = getComputedStyle(ancestor);
-        if (style.visibility === 'hidden' ||
-            style.display === 'none' ||
-            style.opacity === '0' ||
-            style.width === '0' ||
-            style.height === '0') {
-            return false;
-        } else {
-            // It wasn't hidden based on a computed style. See if it's
-            // offscreen:
-            const rect = element.getBoundingClientRect();
-            const frame = element.ownerDocument.defaultView;  // window or iframe
-            if ((rect.right + frame.scrollX < 0) ||
-                (rect.bottom + frame.scrollY < 0)) {
-                return false;
-            }
-        }
+    const rect = element.getBoundingClientRect();
+    if (rect.width === 0 || rect.height === 0) {
+        return false;
+    }
+    const style = getComputedStyle(element);
+    if (style.opacity === '0') {
+        return false;
     }
-    return true;
+    // workaround for https://github.com/w3c/csswg-drafts/issues/4122
+    const scrollX = window.pageXOffset;
+    const scrollY = window.pageYOffset;
+    const absX = rect.x + scrollX;
+    const absY = rect.y + scrollY;
+    window.scrollTo(absX, absY);
```

Comment:
Does this make the page visibly jump? If not now, is there any guaranteed that it won't start visibly jumping in a later Firefox (or other browser)?

Reply:
Admittedly, this workaround is awkward. Though I got some heartening information from Emilio/the web[1], if the user is scrolling while `isVisible` happens to be running, there can be jumpiness. I'll have to revise my approach here.

[1]:
1. I asked Emilio, the graphics/layout engineer who has been helping me analyze alternative approaches:

> I think that work around is ok.

> Javascript needs to run before rendering, the engine cannot simply
interrupt a script and do layout in order to figure out what changed.

> Those calls are sync so I think you're ok.

> My only concern is what happens if the user tries to scroll while that
code is running, since scrolling _is_ async. But I think it should be
fine since we only tell the compositor about the main thread scroll
state after painting.

2. I was reading a [document on best practices](https://developers.google.com/web/fundamentals/performance/rendering/optimize-javascript-execution#reduce_complexity_or_use_web_workers), and it recommends:
> ...if you’re in an animation like scrolling, you should ideally be looking to keep your JavaScript to something in the region of 3-4 ms.

When I look at this implementation’s [profile](https://perfht.ml/2K7Vqhs) in Price Tracker (the Stack Chart, Focus: isVisible), I don’t see any calls exceeding 2 ms, and the average looks to be closer to 1 ms.

<img width="1678" alt="Screen Shot 2019-07-31 at 11 34 04 AM" src="https://user-images.githubusercontent.com/17437436/62244755-f9439b80-b394-11e9-9106-217f901bb3ff.png">